### PR TITLE
Reland [CGData] llvm-cgdata #89884

### PR DESCRIPTION
**Pull request by kyulee-com**
Reland [CGData] llvm-cgdata #89884 using `Opt` instead of `cl`
- Action options are required, `--convert`, `--show`, `--merge`. This was similar to sub-commands previously implemented, but having a prefix `--`.
- `--format` option is added, which specifies `text` or `binary`.

### CODE REVIEW

File: llvm/tools/llvm-cgdata/Opts.td
Comment (MaskRay) [inline llvm/tools/llvm-cgdata/Opts.td:3]: User-facing tools typically use `--foo` (two dashes) style long options. Consider switching to it?

File: llvm/tools/llvm-cgdata/Opts.td
Comment (MaskRay) [inline llvm/tools/llvm-cgdata/Opts.td:23]: For new utility with no concern of backward option compatibility, it's recommended to support only `--`

File: llvm/tools/llvm-cgdata/llvm-cgdata.cpp
Comment (MaskRay) [inline llvm/tools/llvm-cgdata/llvm-cgdata.cpp:75]: StringRef can typically be used if the string is backed by other storage (argv, etc), which also avoids a global destructor.

**Review by aemerson — DISMISSED** [review]
This approach has not been accepted yet. Don't merge until it has been.

File: llvm/tools/llvm-cgdata/Opts.td
Comment (kyulee-com) [reply llvm/tools/llvm-cgdata/Opts.td:3]: Updated by using two dashes for the long options while using single dash for the short (alias) option.

File: llvm/tools/llvm-cgdata/Opts.td
Comment (kyulee-com) [reply llvm/tools/llvm-cgdata/Opts.td]: Yep. Dropped the single dash for this long option.

File: llvm/tools/llvm-cgdata/llvm-cgdata.cpp
Comment (kyulee-com) [reply llvm/tools/llvm-cgdata/llvm-cgdata.cpp]: Use StringRef instead of std::string.

File: llvm/tools/llvm-cgdata/Opts.td
Comment (petrhosek) [inline llvm/tools/llvm-cgdata/Opts.td]: I think we should consider extending https://github.com/llvm/llvm-project/blob/main/llvm/lib/Option/Option.cpp to support subcommands akin to https://github.com/llvm/llvm-project/blob/main/llvm/lib/Support/CommandLine.cpp.

**Comment by kyulee-com** [comment]
> This approach has not been accepted yet. Don't merge until it has been.

Thank you for sharing your concerns about the process for my changes. I understand that it's important to follow the correct process for making decisions in the community. Can you please clarify which specific steps you feel were not followed correctly? I have been seeking feedback through PRs, with the shared RFC (https://discourse.llvm.org/t/rfc-enhanced-machine-outliner-part-2-thinlto-nolto/78753) sitting for months, and have been following the process as I understand it. I am open to working with you to find a solution that aligns with the community's expectations.

File: llvm/tools/llvm-cgdata/Opts.td
Comment (kyulee-com) [reply llvm/tools/llvm-cgdata/Opts.td]: Yes, I agree that it would be beneficial to have subcommand support in `Opt`. However, as discussed in https://github.com/llvm/llvm-project/pull/100077, there seems no explicit support for it yet, and I am not familiar with how to implement this feature. For the time being, I have looked into similar implementations, and tools like `sandcov` group these options as required to mimic subcommand-like behavior, which I'm following -- 
https://github.com/llvm/llvm-project/blob/main/llvm/tools/sancov/Opts.td#L16-L38

File: llvm/lib/CodeGenData/CodeGenDataWriter.cpp
Comment (ellishg) [inline llvm/lib/CodeGenData/CodeGenDataWriter.cpp:70]: Minor NIT:
```suggestion
                       static_cast<const char *>(&Bytes), sizeof(uint64_t));
```

File: llvm/tools/llvm-cgdata/llvm-cgdata.cpp
Comment (ellishg) [inline llvm/tools/llvm-cgdata/llvm-cgdata.cpp:286]: Does this need to be a lambda function? Or can we inline it into the if body below?

File: llvm/tools/llvm-cgdata/llvm-cgdata.cpp
Comment (ellishg) [inline llvm/tools/llvm-cgdata/llvm-cgdata.cpp:307]: Do you expect this to ever have more than one element? Can we use `ArrayRef` or some constant-size container?

File: llvm/lib/CodeGenData/CodeGenDataWriter.cpp
Comment (kyulee-com) [reply llvm/lib/CodeGenData/CodeGenDataWriter.cpp]: Used `reinterpret_cast` instead.

File: llvm/tools/llvm-cgdata/llvm-cgdata.cpp
Comment (kyulee-com) [reply llvm/tools/llvm-cgdata/llvm-cgdata.cpp]: Inlined the code.

File: llvm/tools/llvm-cgdata/llvm-cgdata.cpp
Comment (kyulee-com) [reply llvm/tools/llvm-cgdata/llvm-cgdata.cpp]: Loop thru Args to ensure only one action arg, instead of using `SmallVector`.

File: llvm/lib/CodeGenData/CodeGenDataWriter.cpp
Comment (MaskRay) [inline llvm/lib/CodeGenData/CodeGenDataWriter.cpp:21]: In .cpp files, we don't use `namespace xxx { }` https://llvm.org/docs/CodingStandards.html#use-namespace-qualifiers-to-implement-previously-declared-functions

**Comment by MaskRay** [comment]
Related patches have introduced `llvm/lib/CodeGenData` and `llvm/include/llvm/CodeGenData/`. Adding new directories under `llvm/lib` and `llvm/include/llvm` require more discussions.

I think you need to ask more folks about their opinions, my intuition is that reusing `llvm/lib/Object` might be a choice.

File: llvm/lib/CodeGenData/CodeGenDataWriter.cpp
Comment (kyulee-com) [reply llvm/lib/CodeGenData/CodeGenDataWriter.cpp]: Moved the type definitions to the `.h` file and removed the `namespace llvm {` from the `.cpp` file.

**Comment by kyulee-com** [comment]
> Related patches have introduced `llvm/lib/CodeGenData` and `llvm/include/llvm/CodeGenData/`. Adding new directories under `llvm/lib` and `llvm/include/llvm` require more discussions.
> 
> I think you need to ask more folks about their opinions, my intuition is that reusing `llvm/lib/Object` might be a choice. The `.cpp` files seem to use Support and Object, and therefore need to live at a layer higher than LLVMObject per https://llvm.org/docs/CodingStandards.html#library-layering

Thank you for the guidance on layering within lib and include. Based on the conventions from the link you provided, placing `CodeGenData` files under `llvm/lib/Object` is an option. However, to avoid clutter and maintain distinct functionality, a separate `CodeGenData` directory might be preferable, similar to the `ProfileData` setup which also uses `Support` and `Object`.

I'm open to other opinions on this matter. Could you please add more reviewers relevant to this discussion? If the consensus favors a different placement, I will create a separate PR (NFC) to accommodate that change, instead of making changes in place here.

**Comment by MaskRay** [comment]
> > Related patches have introduced `llvm/lib/CodeGenData` and `llvm/include/llvm/CodeGenData/`. Adding new directories under `llvm/lib` and `llvm/include/llvm` require more discussions.
> > I think you need to ask more folks about their opinions, my intuition is that reusing `llvm/lib/Object` might be a choice. The `.cpp` files seem to use Support and Object, and therefore need to live at a layer higher than LLVMObject per [llvm.org/docs/CodingStandards.html#library-layering](https://llvm.org/docs/CodingStandards.html#library-layering)
> 
> Thank you for the guidance on layering within lib and include. Based on the conventions from the link you provided, placing `CodeGenData` files under `llvm/lib/Object` is an option. However, to avoid clutter and maintain distinct functionality, a separate `CodeGenData` directory might be preferable, similar to the `ProfileData` setup which also uses `Support` and `Object`.
> 
> I'm open to other opinions on this matter. Could you please add more reviewers relevant to this discussion? If the consensus favors a different placement, I will create a separate PR (NFC) to accommodate that change, instead of making changes in place here.

Beside existing reviewers, @efriedma-quic @nikic @RKSimon Do you have suggestions on the directory name? (currently `llvm/lib/CodeGenData/`, which I feel a bit confusing when `llvm/lib/CodeGenTypes/` is present)

**Comment by RKSimon** [comment]
Why isn't this just put inside llvm/tools/llvm-cgdata/lib ? See llvm/tools/llvm-exegesis for an example of the lib being separated from the tool.

**Comment by kyulee-com** [comment]
@aemerson Thank you for reviewing RFC and unblocking this PR!

> currently llvm/lib/CodeGenData/, which I feel a bit confusing when llvm/lib/CodeGenTypes/ is present

@MaskRay What about `CGData` or something else to avoid this confusion? This library name also seems more consistent to the tool name, `llvm-cgdata`.

**Comment by MaskRay** [comment]
> @aemerson Thank you for reviewing RFC and unblocking this PR!
> 
> > currently llvm/lib/CodeGenData/, which I feel a bit confusing when llvm/lib/CodeGenTypes/ is present
> 
> @MaskRay What about `CGData` or something else to avoid this confusion? This library name also seems more consistent to the tool name, `llvm-cgdata`.

`llvm/lib/CGData` looks good to me (analogy: ObjCopy for llvm-objcopy).

**Comment by kyulee-com** [comment]
I think I've addressed all the comments. Can someone review the changes again if everything looks okay?

**Review by thevinster — APPROVED** [review]
Seems like everything has been addressed and changes lgtm.

File: llvm/unittests/CGData/CMakeLists.txt
Comment (nico) [inline llvm/unittests/CGData/CMakeLists.txt:7]: Since the directory got renamed, shouldn't the exe be called `CGDataTests` instead of `CodeGenDataTests` too?

Comment (kyulee-com) [reply llvm/unittests/CGData/CMakeLists.txt:7]: Yeah. Thanks for catching this! Will follow this up soon.

Comment (kyulee-com) [reply llvm/unittests/CGData/CMakeLists.txt:7]: https://github.com/llvm/llvm-project/pull/105463

**Comment by jakeegan** [comment]
Hi, the tests are failing on the AIX bot https://lab.llvm.org/buildbot/#/builders/64/builds/745 because `-i` is not supported with sed on AIX. Would you be able to adapt the test?

**Comment by kyulee-com** [comment]
> Hi, the tests are failing on the AIX bot https://lab.llvm.org/buildbot/#/builders/64/builds/745 because `-i` is not supported with sed on AIX. Would you be able to adapt the test?

Thanks for reporting this issue! I fixed the tests without using options -- https://github.com/llvm/llvm-project/pull/105546. Hope it addresses this issue.

**Pull request merged**
